docs/probes.md

```diff
@@ -0,0 +1,130 @@
+# Readiness and LivenessProbes
+
+The VMI spec allows setting `livenessProbe` and `readinessProbe` which translate to the same field on the resulting pod running the VM.
+
+## Exec Probes
+
+One option to probe the VM is by running commands on it and determine the ready/live state based on it's success.
+
+Compared to pods, we need a way to execute the command inside the actual VM and not the pod.
+To do that, we rely on the qemu-guest-agent to be available inside the VM.
+
+A command supplied to an exec probe, will be wrapped by `virt-probe` in the operator and forwarded to the guest.
+The probe will fail if the agent is not available.
+
+Make sure to provide enough delay and failureThreshold for the VM and the agent to be online.
+
+### Example
+
+**Note:** The Fedora image used in this example does not have the qemu-guest-agent available by default. We need to install it through the console until this is resolved.
+
+1. Create VM manifest
+
+```yaml
+# /tmp/probe-test.vm.yaml
+apiVersion: kubevirt.io/v1alpha3
+kind: VirtualMachine
+metadata:
+  name: probe-test
+  namespace: default
+  labels:
+    app: probe-test
+    vm.kubevirt.io/name: probe-test
+    kubevirt.io/domain: probe-test
+spec:
+  running: false
+  template:
+    metadata:
+      labels:
+        vm.kubevirt.io/name: probe-test
+        kubevirt.io/domain: probe-test
+    spec:
+      readinessProbe:
+        exec:
+          command:
+          - cat
+          - /tmp/ready.txt
+        failureThreshold: 10
+        initialDelaySeconds: 120
+        periodSeconds: 10
+        timeoutSeconds: 1
```

Comment:
as an example, i'd give this a slightly higher timeout. maybe 5 seconds? people will copy and paste this

Reply:
We should make sure the probe is safe.
Right now before v1.20 Kubernetes the timeoutSeconds would not get used correctly.
https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#configure-probes

The note on that link also plays into your question regarding memory overhead as in the worst case it might cause the problem mentioned there:
```
Note:
Before Kubernetes 1.20, the field timeoutSeconds was not respected for exec probes: probes continued running indefinitely, even past their configured deadline, until a result was returned.

This defect was corrected in Kubernetes v1.20. You may have been relying on the previous behavior, even without realizing it, as the default timeout is 1 second. As a cluster administrator, you can disable the feature gate ExecProbeTimeout (set it to false) on each kubelet to restore the behavior from older versions, then remove that override once all the exec probes in the cluster have a timeoutSeconds value set.
If you have pods that are impacted from the default 1 second timeout, you should update their probe timeout so that you're ready for the eventual removal of that feature gate.

With the fix of the defect, for exec probes, on Kubernetes 1.20+ with the dockershim container runtime, the process inside the container may keep running even after probe returned failure because of the timeout.

Caution: Incorrect implementation of readiness probes may result in an ever growing number of processes in the container, and resource starvation if this is left unchecked.

```

Reply:
ha, that would be bad.

My concern about memory is due to us getting OOM killed due to components within the pod tipping the pod over the requested/limited memory requirements.  we'lve had issues wiht that in the past, which is why we calculated memory overhead of our components in virt-launcher pod

Reply:
I'm building a small fake cmdserver right now so I can profile the virt-probe command to see how it does memory wise.